src/librustdoc/clean/types.rs

```diff
@@ -57,11 +57,18 @@ crate struct Crate {
     crate primitives: Vec<(DefId, PrimitiveType)>,
     // These are later on moved into `CACHEKEY`, leaving the map empty.
     // Only here so that they can be filtered through the rustdoc passes.
-    crate external_traits: Rc<RefCell<FxHashMap<DefId, Trait>>>,
+    crate external_traits: Rc<RefCell<FxHashMap<DefId, TraitWithExtraInfo>>>,
     crate masked_crates: FxHashSet<CrateNum>,
     crate collapsed: bool,
 }
 
+/// This struct is used to wrap additional information added by rustdoc on a `trait` item.
+#[derive(Clone, Debug)]
+crate struct TraitWithExtraInfo {
+    crate trait_: Trait,
+    crate is_spotlight: bool,
+}
```

Comment:
The whole point of this PR is to remove the field from `Trait`, isn't it? I feel a bit confused as well.

Reply:
Oh right, that too. Why not calculate this on demand?

Reply:
https://github.com/rust-lang/rust/pull/80914#pullrequestreview-578522364

Maybe we could use some sort of on-demand system with caching? Hmm... that makes me thinking of queries—can we define queries in rustdoc?

I wonder why this is having a positive perf impact given that it's still calculating ahead-of-time... I must be missing something.

Reply:
Queries must be declared in rustc_middle, which means any types they use must be available that earlier. So we can't have a query that returns `clean::Trait`, but it's feasible to add one for `implement_doc_spotlight(DefId) -> bool`. That would need sign-off from a compiler-team member though.

Reply:
Should I add a FIXME so we can do it in another PR?

Reply:
Well, I'm not sure what the point of this PR is if you're still storing `is_spotlight` in memory. Why not calculate it everywhere it's used?

Reply:
The point is to make `ItemKind` looks as close as possible to the compiler type so that it makes the switch simpler (which is why I make all these "small" PRs).

Reply:
> Queries must be declared in rustc_middle, which means any types they use must be available that early. So we can't have a query that returns `clean::Trait`, but it's feasible to add one for `implement_doc_spotlight(DefId) -> bool`. That would need sign-off from a compiler-team member though.

Better yet, what about adding `has_doc_flag(DefId, Symbol) -> bool` so it's more general-purpose?

Reply:
> The point is to make ItemKind looks as close as possible to the compiler type so that it makes the switch simpler (which is why I make all these "small" PRs).

Could you add that to the PR description?